README
Insilicut - in silico cutting of genomic DNA with restriction enzyme
====================================================================

This directory contains insilicut, a minimal package reporting fragments when
cutting in silico genomic DNA with a restriction enzyme. It is potentially 
useful when planning a genotyping-by-sequencing (GBS) experiment (also known 
as restriction-assisted DNA sequencing, RAD-seq).

This package (copyright INRA) is free software, you can redistribute it 
and/or modify it under the terms of the GNU General Public License (GPL-3).

The GNU General Public License does not permit this software to be
redistributed in proprietary programs.

This library is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

The code is hosted online: https://github.com/timflutre/insilicut


Installation
============

Insilicut doesn't need any compilation as it only contains scripts. However, 
it still has a few requirements:
* Bash and Awk;
* PaTMaN (https://bioinf.eva.mpg.de/patman/);
* Python (version >= 2.7 or 3.1);
* R (optional, version >= 2.15).

First, download the package, for instance:
    wget --no-check-certificate https://github.com/timflutre/insilicut/archive/master.zip
    unzip master; rm -f master; cd insilicut-master

Then, run the tests:
    make check

Finally, install the package:
    make install

By default everything is installed in a directory bin/ in ${HOME}. If the 
bin/ directory does not exist, it will be created. To install elsewhere, use 
the option INSTALL, for instance:
    make install INSTALL="/usr/local"


Documentation
=============

You can access the manual in command-line via:
    man ./doc/insilicut.man

To re-generate the manual in PDF format, type:
    make pdf


NEWS
2015-06-30 v1.0.1
=================

minor fixes


2014-04-27 v1.0
===============

first public release


tests/test_func.bash
#!/usr/bin/env bash

# Aim: launch a functional test for the insilicut package
# Copyright (C) 2014 Institut National de la Recherche Agronomique (INRA)
# License: GPL-3+
# Author: Timothée Flutre

progVersion="1.0"

# Display the help on stdout.
# The format complies with help2man (http://www.gnu.org/s/help2man)
function help () {
  msg="\`${0##*/}' launches a functional test for the insilicut package.\n"
  msg+="\n"
  msg+="Usage: ${0##*/} [OPTIONS] ...\n"
  msg+="\n"
  msg+="Options:\n"
  msg+="  -h, --help\tdisplay the help and exit\n"
  msg+="  -V, --version\toutput version information and exit\n"
  msg+="  -v, --verbose\tverbosity level (0/default=1/2/3)\n"
  msg+="  -i, --p2i\tabsolute path to the insilicut directory\n"
  msg+="  -n, --noclean\tkeep temporary directory with all files\n"
  msg+="\n"
  msg+="Report bugs to <timothee.flutre@supagro.inra.fr>."
  echo -e "$msg"
}

# Display version and license information on stdout.
function version () {
  msg="${0##*/} ${progVersion}\n"
  msg+="\n"
  msg+="Copyright (C) 2014 Institut National de la Recherche Agronomique (INRA).\n"
  msg+="License GPL-3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n"
  msg+="This is free software; see the source for copying conditions. There is NO\n"
  msg+="warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
  msg+="\n"
  msg+="Written by Timothée Flutre."
  echo -e "$msg"
}

# http://www.linuxjournal.com/content/use-date-command-measure-elapsed-time
function timer () {
  if [[ $# -eq 0 ]]; then
    echo $(date '+%s')
  else
    local startRawTime=$1
    endRawTime=$(date '+%s')
    if [[ -z "$startRawTime" ]]; then startRawTime=$endRawTime; fi
    elapsed=$((endRawTime - startRawTime)) # in sec
    nbDays=$((elapsed / 86400))
    nbHours=$(((elapsed / 3600) % 24))
    nbMins=$(((elapsed / 60) % 60))
    nbSecs=$((elapsed % 60))
    printf "%01dd %01dh %01dm %01ds" $nbDays $nbHours $nbMins $nbSecs
  fi
}

# Parse the command-line arguments.
# http://stackoverflow.com/a/4300224/597069
function parseCmdLine () {
  getopt -T > /dev/null # portability check (say, Linux or Mac OS?)
  if [ $? -eq 4 ]; then # GNU enhanced getopt is available
	  TEMP=`getopt -o hVv:i:n -l help,version,verbose:,p2i:,noclean \
        -n "$0" -- "$@"`
  else # original getopt is available (no long options, whitespace, sorting)
	  TEMP=`getopt hVv:i:n "$@"`
  fi
  if [ $? -ne 0 ]; then
	  echo "ERROR: "$(which getopt)" failed"
	  getopt -T > /dev/null
	  if [ $? -ne 4 ]; then
	    echo "did you use long options? they are not handled \
on your system, use -h for help"
	  fi
	  exit 2
  fi
  eval set -- "$TEMP"
  while [ $# -gt 0 ]; do
    case "$1" in
      -h | --help) help; exit 0; shift;;
      -V | --version) version; exit 0; shift;;
      -v | --verbose) verbose=$2; shift 2;;
      -i | --p2i) pathToInsilicut=$2; shift 2;;
	    -n | --noclean) clean=false; shift;;
      --) shift; break;;
      *) echo "ERROR: options parsing failed, use -h for help"; exit 1;;
    esac
  done
  if [ -z "${pathToInsilicut}" ]; then
    echo -e "ERROR: missing compulsory option --p2i\n"
    help
    exit 1
  fi
  if [ ! -d "${pathToInsilicut}" ]; then
    echo -e "ERROR: can't find directory ${pathToInsilicut}\n"
    help
    exit 1
  fi
}

function run () {
  cwd=$(pwd)
  cd "${pathToInsilicut}/tests"
  
  # step 1 ------------------------------------------------------------------
  if [ $verbose -gt "0" ]; then
	  echo -e "prepare input data..."
  fi
  if [ ! -f "${pathToInsilicut}/tests/TAIR10_chr1.fas" ]; then
	  if [ $verbose -gt "0" ]; then
	    echo -e "download chromosome 1 from A. thaliana..."
	  fi
	  wget ftp://ftp.arabidopsis.org/home/tair/Sequences/whole_chromosomes/TAIR10_chr1.fas
  fi
  echo -e ">ApeKI\nGCWGC" > ApeKI.fa
  
  # step 2 ------------------------------------------------------------------
  uniqId=$$ # process ID
  testDir=tmp_test_${uniqId}
  rm -rf ${testDir}
  mkdir ${testDir}
  cd ${testDir}
  if [ $verbose -gt "0" ]; then echo "temp dir: "$(pwd); fi
  
  # step 3 ------------------------------------------------------------------
  if [ $verbose -gt "0" ]; then
	  echo -e "launch insilicut..."
  fi
  ${pathToInsilicut}/scripts/insilicut.bash --gf ../TAIR10_chr1.fas \
	                  --gn Atha --ef ../ApeKI.fa --en ApeKI -v $(expr ${verbose} - 1) \
	                  --p2i ${pathToInsilicut}
  
  # step 4 ------------------------------------------------------------------
  if [ $verbose -gt "0" ]; then
	  echo -e "check outputs..."
  fi
  if [ $(zcat out_patman_Atha_ApeKI_e-0_g-0_a_frags_s-100_S-300.bed.gz | wc -l) != 5968 ]; then
	  echo -e "test failed!"
	  exit 1
  else
	  echo -e "test passed!"
  fi
  
  # step 5 ------------------------------------------------------------------
  cd ${cwd}
  if $clean; then rm -rf "${pathToInsilicut}/tests/${testDir}"; fi
}

verbose=1
pathToInsilicut=""
clean=true
parseCmdLine "$@"

if [ $verbose -gt "0" ]; then
  startTime=$(timer)
  msg="START ${0##*/} $(date +"%Y-%m-%d") $(date +"%H:%M:%S")"
  msg+="\ncmd-line: $0 "$@ # comment if an option takes a glob as argument
  msg+="\ncwd: $(pwd)"
  echo -e $msg
fi

run pathToInsilicut clean

if [ $verbose -gt "0" ]; then
  msg="END ${0##*/} $(date +"%Y-%m-%d") $(date +"%H:%M:%S")"
  msg+=" ($(timer startTime))"
  echo $msg
fi


scripts/insilicut_extract_fragments.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Aim: extract non-overlapping fragments in a given size range from a BED file
# containing restriction sites from in silico RAD-seq
# Copyright (C) 2014 Institut National de la Recherche Agronomique (INRA)
# License: GPL-3+
# Author: Timothée Flutre

# to allow code to work with Python 2 and 3
from __future__ import print_function   # print is a function in python3
from __future__ import unicode_literals # avoid adding "u" to each string
from __future__ import division # avoid writing float(x) when dividing by x

import sys
import os
import getopt
import time
import datetime
from subprocess import Popen, PIPE
import math
import gzip
import copy
import numpy as np
# import scipy as sp

if sys.version_info[0] == 2:
    if sys.version_info[1] < 7:
        msg = "ERROR: Python should be in version 2.7 or higher"
        sys.stderr.write("%s\n\n" % msg)
        sys.exit(1)
        
progVersion = "1.1.1" # http://semver.org/


class Bed(object):
    def __init__(self, chrom, start, end, name, score=None, strand=None):
        self.chrom = chrom
        self.start = int(start) # 0-based
        self.end = int(end)
        self.name = name
        self.score = int(score) if score != None else None
        self.strand = strand
    def __len__(self):
        return self.end - self.start
    def __repr__(self):
        txt = "%s\t%i\t%i\t%s" % (self.chrom, self.start, self.end, self.name)
        if self.score != None:
            txt += "\t%i" % self.score
        if self.strand != None:
            txt += "\t%s" % self.strand
        return txt
        
        
class ExtractFragments(object):
    
    def __init__(self):
        self.verbose = 1
        self.inBedFile = ""
        self.minFragSize = 1
        self.maxFragSize = 500
        self.outBedPrefix = ""
        
        
    def help(self):
        """
        Display the help on stdout.
        
        The format complies with help2man (http://www.gnu.org/s/help2man)
        """
        msg = "`%s' extracts non-overlapping fragments in a given size range\nfrom a BED file containing restriction sites from in silico RAD-seq\n" % os.path.basename(sys.argv[0])
        msg += "\n"
        msg += "Usage: %s [OPTIONS] ...\n" % os.path.basename(sys.argv[0])
        msg += "\n"
        msg += "Options:\n"
        msg += " -h, --help\tdisplay the help and exit\n"
        msg += " -V, --version\toutput version information and exit\n"
        msg += " -v, --verbose\tverbosity level (0/default=1/2/3)\n"
        msg += " -i\t\tinput file with the cut sites (BED format, gzipped)\n"
        msg += "\t\tcoordinates should correspond to the whole motif (e.g. 7-11 instead of 9)\n"
        msg += " -s\t\tmin size of the fragments to select (default=1)\n"
        msg += " -S\t\tmax size of the fragments to select (default=500)\n"
        msg += "\n"
        msg += "Remarks:\n"
        msg += " two output files are written, one with all fragments and one with the size-selected fragments"
        msg += "\n"
        msg += "Report bugs to <timothee.flutre@supagro.inra.fr>."
        print(msg); sys.stdout.flush()
        
        
    def version(self):
        """
        Display version and license information on stdout.
        """
        msg = "%s %s\n" % (os.path.basename(sys.argv[0]), progVersion)
        msg += "\n"
        msg += "Copyright (C) 2014 Institut National de la Recherche Agronomique (INRA).\n"
        msg += "License GPL-3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n"
        msg += "This is free software; see the source for copying conditions. There is NO\n"
        msg += "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
        msg += "\n"
        msg += "Written by Timothée Flutre."
        print(msg.encode("utf8")); sys.stdout.flush()
        
        
    def setAttributesFromCmdLine(self):
        """
        Parse the command-line arguments.
        """
        try:
            opts, args = getopt.getopt( sys.argv[1:], "hVv:i:s:S:",
                                        ["help", "version", "verbose="])
        except getopt.GetoptError, err:
            sys.stderr.write("%s\n" % str(err))
            self.help()
            sys.exit(2)
        for o, a in opts:
            if o == "-h" or o == "--help":
                self.help()
                sys.exit(0)
            elif o == "-V" or o == "--version":
                self.version()
                sys.exit(0)
            elif o == "-v" or o == "--verbose":
                self.verbose = int(a)
            elif o == "-i":
                self.inBedFile = a
            elif o == "-s":
                self.minFragSize = int(a)
            elif o == "-S":
                self.maxFragSize = int(a)
            else:
                assert False, "unhandled option"
                
                
    def checkAttributes(self):
        """
        Check the values of the command-line parameters.
        """
        if self.inBedFile == "":
            msg = "ERROR: missing compulsory option -i"
            sys.stderr.write("%s\n\n" % msg)
            self.help()
            sys.exit(1)
        if not os.path.exists(self.inBedFile):
            msg = "ERROR: can't find '%s'" % self.inBedFile
            sys.stderr.write("%s\n\n" % msg)
            self.help()
            sys.exit(1)
        self.outBedPrefix = self.inBedFile.split(".bed.gz")[0]
        
        
    def loadBedFile(self):
        if self.verbose > 0:
            sys.stdout.write("read file %s ..." % self.inBedFile)
            sys.stdout.flush()
        inBedH = gzip.open(self.inBedFile)
        lines = inBedH.readlines()
        inBedH.close()
        if self.verbose > 0:
            sys.stdout.write(" done\n")
        return lines
    
    
    def parseLines(self, lines):
        dChr2sites = {}
        for line in lines:
            tokens = line.rstrip().split("\t")
            if tokens[0] not in dChr2sites:
                dChr2sites[tokens[0]] = []
            dChr2sites[tokens[0]].append(Bed(tokens[0], tokens[1], tokens[2],
                                             tokens[3]))
        return dChr2sites
        
        
    def getNonOverlappingFragments(self, dChr2sites):
        lFrags = []
        if self.verbose > 0:
            msg = "get non-overlapping fragments ..."
            print(msg); sys.stdout.flush()
        chroms = dChr2sites.keys()
        chroms.sort()
        
        for chrom in chroms:
            if self.verbose > 0:
                msg = "%s: %i cuts" % (chrom, len(dChr2sites[chrom]))
            dChr2sites[chrom].sort(key=lambda x: x.start)
            nbFrags = 0
            # lFrags.append(Bed(chrom, 0, dChr2sites[chrom][0].start,
            #                   dChr2sites[chrom][0].name))
            nbFrags += 1
            for i in xrange(1, len(dChr2sites[chrom])):
                if dChr2sites[chrom][i-1].end < dChr2sites[chrom][i].start:
                    lFrags.append(Bed(chrom, dChr2sites[chrom][i-1].end,
                                      dChr2sites[chrom][i].start,
                                      "%s_%s" % (dChr2sites[chrom][i-1].name, dChr2sites[chrom][i].name)))
                    nbFrags += 1
            if self.verbose > 0:
                msg += " and %i fragments" % nbFrags
                print(msg); sys.stdout.flush()
                
        if self.verbose > 0:        
            lenFrags = np.array([len(x) for x in lFrags], dtype=int)
            print("nb of fragments: %i (mean-len=%.2f std-err=%f std-dev=%.2f min=%i Q25=%.2f med=%i Q75=%.2f max=%i)" \
                  % (len(lFrags), np.mean(lenFrags),
                     np.std(lenFrags)/math.sqrt(len(lFrags)),
                     np.std(lenFrags),
                     np.min(lenFrags),
                     np.percentile(lenFrags, 25),
                     np.median(lenFrags),
                     np.percentile(lenFrags, 75),
                     np.max(lenFrags)))
        return lFrags
        
        
    def saveFragments(self, lFrags):
        if self.verbose > 0:        
            print("save fragments of the good size ...")
            sys.stdout.flush()
        outBedAllFile = "%s_frags.bed.gz" % self.outBedPrefix
        outBedAllH = gzip.open(outBedAllFile, "w")
        outBedSelFile = "%s_frags_s-%i_S-%i.bed.gz" % (self.outBedPrefix,
                                                       self.minFragSize,
                                                       self.maxFragSize)
        outBedSelH = gzip.open(outBedSelFile, "w")
        nbKeptFrags = 0
        for frag in lFrags:
            outBedAllH.write("%s\n" % frag)
            if len(frag) >= self.minFragSize and \
               len(frag) <= self.maxFragSize:
                outBedSelH.write("%s\n" % frag)
                nbKeptFrags += 1
        outBedAllH.close()
        outBedSelH.close()
        if self.verbose > 0:        
            print("nb of kept fragments: %i" % nbKeptFrags)
            
            
    def run(self):
        lines = self.loadBedFile()
        dChr2sites = self.parseLines(lines)
        lFrags = self.getNonOverlappingFragments(dChr2sites)
        self.saveFragments(lFrags)
        
        
if __name__ == "__main__":
    i = ExtractFragments()
    
    i.setAttributesFromCmdLine()
    
    i.checkAttributes()
    
    if i.verbose > 0:
        startTime = time.time()
        msg = "START %s %s" % (os.path.basename(sys.argv[0]),
                               time.strftime("%Y-%m-%d %H:%M:%S"))
        msg += "\ncmd-line: %s" % ' '.join(sys.argv)
        msg += "\ncwd: %s" % os.getcwd()
        print(msg); sys.stdout.flush()
        
    i.run()
    
    if i.verbose > 0:
        msg = "END %s %s" % (os.path.basename(sys.argv[0]),
                             time.strftime("%Y-%m-%d %H:%M:%S"))
        endTime = time.time()
        runLength = datetime.timedelta(seconds=
                                       math.floor(endTime - startTime))
        msg += " (%s" % str(runLength)
        if "linux" in sys.platform:
            p = Popen(["grep", "VmHWM", "/proc/%s/status" % os.getpid()],
                      shell=False, stdout=PIPE).communicate()
            maxMem = p[0].split()[1]
            msg += "; %s kB)" % maxMem
        else:
            msg += ")"
        print(msg); sys.stdout.flush()


scripts/insilicut.bash
#!/usr/bin/env bash

# Aim: report fragments when cutting genomic DNA with a restriction enzyme
# Copyright (C) 2014-2015 Institut National de la Recherche Agronomique (INRA)
# License: GPL-3+
# Persons: Timothée Flutre [cre,aut]
# Versioning: https://github.com/timflutre/insilicut

progVersion="1.1.1" # http://semver.org/

# Display the help on stdout.
# The format complies with help2man (http://www.gnu.org/s/help2man)
function help () {
  msg="\`${0##*/}' reports fragments when cutting genomic DNA with a restriction enzyme.\n"
  msg+="\n"
  msg+="Usage: ${0##*/} [OPTIONS] ...\n"
  msg+="\n"
  msg+="Options:\n"
  msg+="  -h, --help\tdisplay the help and exit\n"
  msg+="  -V, --version\toutput version information and exit\n"
  msg+="  -v, --verbose\tverbosity level (0/default=1/2/3)\n"
  msg+="      --gf\tpath to the file containing the genomic DNA (fasta format)\n"
  msg+="      --gn\tname of the genomic DNA (e.g. 'Athaliana')\n"
  msg+="      --ef\tpath to the file containing the restriction enzyme (fasta format)\n"
  msg+="      --en\tname of the enzyme (e.g. 'ApeKI')\n"
  msg+="      --ls\tlower bound on fragments size (default=100)\n"
  msg+="      --us\tupper bound on fragments size (default=300)\n"
  msg+="      --fullh\tuse full fasta header (e.g. keep \"chr1 Atha\" instead of \"chr1\")\n"
  msg+="      --clean\tremove temporary files\n"
  msg+="      --p2i\tabsolute path to the insilicut directory (default="")\n"
  msg+="\t\tused for testing purposes only (e.g. in 'make check')\n"
  msg+="\n"
  msg+="Examples:\n"
  msg+="  ${0##*/} --gf Athaliana_genome.fa --gn Athaliana --ef ApeKI.fa --en ApeKI\n"
  msg+="\n"
  msg+="Remarks:\n"
  msg+="  if R is installed, a histogram of fragments size is also produced\n"
  msg+="\n"
  msg+="Report bugs to <timothee.flutre@supagro.inra.fr>."
  echo -e "$msg"
}

# Display version and license information on stdout.
# The person roles comply with R's guidelines (The R Journal Vol. 4/1, June 2012).
function version () {
  msg="${0##*/} ${progVersion}\n"
  msg+="\n"
  msg+="Copyright (C) 2014-2015 Institut National de la Recherche Agronomique (INRA).\n"
  msg+="License GPL-3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>\n"
  msg+="This is free software; see the source for copying conditions. There is NO\n"
  msg+="warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"
  msg+="\n"
  msg+="Written by Timothée Flutre [cre,aut]."
  echo -e "$msg"
}

# http://www.linuxjournal.com/content/use-date-command-measure-elapsed-time
function timer () {
  if [[ $# -eq 0 ]]; then
    echo $(date '+%s')
  else
    local startRawTime=$1
    endRawTime=$(date '+%s')
    if [[ -z "$startRawTime" ]]; then startRawTime=$endRawTime; fi
    elapsed=$((endRawTime - startRawTime)) # in sec
    nbDays=$((elapsed / 86400))
    nbHours=$(((elapsed / 3600) % 24))
    nbMins=$(((elapsed / 60) % 60))
    nbSecs=$((elapsed % 60))
    printf "%01dd %01dh %01dm %01ds" $nbDays $nbHours $nbMins $nbSecs
  fi
}

# Parse the command-line arguments.
# http://stackoverflow.com/a/4300224/597069
function parseCmdLine () {
  getopt -T > /dev/null # portability check (say, Linux or Mac OS?)
  if [ $? -eq 4 ]; then # GNU enhanced getopt is available
	  TEMP=`getopt -o hVv:g:e: -l help,version,verbose:,gf:,gn:,ef:,en:,ls:,us:,fullh,clean,p2i: -n "$0" -- "$@"`
  else # original getopt is available (no long options, whitespace, sorting)
	  TEMP=`getopt hVv: "$@"`
  fi
  if [ $? -ne 0 ]; then
	  echo "ERROR: "$(which getopt)" failed"
	  getopt -T > /dev/null
	  if [ $? -ne 4 ]; then
	    echo "did you use long options? they are not handled \
on your system, use -h for help"
	  fi
	  exit 2
  fi
  eval set -- "$TEMP"
  while [ $# -gt 0 ]; do
    case "$1" in
      -h | --help) help; exit 0; shift;;
      -V | --version) version; exit 0; shift;;
      -v | --verbose) verbose=$2; shift 2;;
      --gf) genomeFile=$2; shift 2;;
	    --gn) genomeName=$2; shift 2;;
      --ef) enzymeFile=$2; shift 2;;
	    --en) enzymeName=$2; shift 2;;
	    --ls) lowerSize=$2; shift 2;;
	    --us) upperSize=$2; shift 2;;
      --fullh) useFullHeader="yes"; shift;;
	    --clean) cleanTmp=true; shift;;
	    --p2i) pathToInsilicut=$2; shift 2;;
      --) shift; break;;
      *) echo "ERROR: options parsing failed, use -h for help"; exit 1;;
    esac
  done
  if [ -z "${genomeFile}" ]; then
    echo -e "ERROR: missing compulsory option --gf\n"
    help
    exit 1
  fi
  if [ ! -f "${genomeFile}" ]; then
    echo -e "ERROR: can't find file ${genomeFile}\n"
    help
    exit 1
  fi
  if [ -z "${genomeName}" ]; then
    echo -e "ERROR: missing compulsory option --gn\n"
    help
    exit 1
  fi
  if [ -z "${enzymeFile}" ]; then
    echo -e "ERROR: missing compulsory option --ef\n"
    help
    exit 1
  fi
  if [ ! -f "${enzymeFile}" ]; then
    echo -e "ERROR: can't find file ${enzymeFile}\n"
    help
    exit 1
  fi
  if [ -z "${enzymeName}" ]; then
    echo -e "ERROR: missing compulsory option --en\n"
    help
    exit 1
  fi
  if ! hash patman 2>/dev/null; then
	  echo -e "ERROR: can't find program 'patman' (https://bioinf.eva.mpg.de/patman/) in PATH\n"
	  exit 1
  fi
  if [ -z "${pathToInsilicut}" ]; then
	  if ! hash insilicut_extract_fragments.py 2>/dev/null; then
	    echo -e "ERROR: can't find program 'insilicut_extract_fragments.py' in PATH\n"
	    exit 1
	  fi
  else
	  pathToInsilicut="${pathToInsilicut}/scripts/"
	  if [ ! -f "${pathToInsilicut}insilicut_extract_fragments.py" ]; then
	    echo -e "ERROR: can't find program 'insilicut_extract_fragments.py' in package\n"
	    exit 1
	  fi
  fi
}

function run () {
  # step 1 ------------------------------------------------------------------
  if [ $verbose -gt "0" ]; then
	  echo -e "find all cut sites with patman..."
  fi
  tmpPrefix=out_patman_${genomeName}_${enzymeName}_e-0_g-0_a
  if [ ! -f "${tmpPrefix}.txt" ]; then
	  patman -P ${enzymeFile} -D ${genomeFile} -e 0 -g 0 -o ${tmpPrefix}.txt -a -s
  fi
  nbCutSites=$(wc -l < ${tmpPrefix}.txt)
  if [ $verbose -gt "0" ]; then
	  echo -e "nb of cut sites: "$nbCutSites
  fi
  
  if [ $nbCutSites -eq "0" ]; then
	  if $cleanTmp; then
	    rm -f ${tmpPrefix}.txt
	  fi
  else
	  # step 2 ------------------------------------------------------------------
	  if [ $verbose -gt "0" ]; then
	    echo -e "convert output into BED format (with AWK)..."
	  fi
	  if [ ! -f "${tmpPrefix}.bed.gz" ]; then
      if [ "${useFullHeader}" == "yes" ]; then
	      cat ${tmpPrefix}.txt | awk -F "\t" '{print $1"\t"$3-1"\t"$4"\t"$2"\t1000\t"$5}' | gzip > ${tmpPrefix}.bed.gz
      else
        cat ${tmpPrefix}.txt | awk -F "\t" '{split($1,a," "); print a[1]"\t"$3-1"\t"$4"\t"$2"\t1000\t"$5}' | gzip > ${tmpPrefix}.bed.gz
      fi
	  fi
	  if $cleanTmp; then
	    rm -f ${tmpPrefix}.txt
	  fi
	  
	  # step 3 ------------------------------------------------------------------
	  if [ $verbose -gt "0" ]; then
	    echo -e "extract fragments (with Python)..."
	  fi
	  if [ ! -f "${tmpPrefix}_frags.bed.gz" ]; then
	    ${pathToInsilicut}insilicut_extract_fragments.py -i ${tmpPrefix}.bed.gz -s ${lowerSize} -S ${upperSize} -v ${verbose} >& stdout_extract_fragments_${genomeName}_${enzymeName}_e-0_g-0_a_s-${lowerSize}_S-${upperSize}.txt
	    if [ $verbose -gt "0" ]; then
		    echo -e $(grep "nb of kept fragments" stdout_extract_fragments_${genomeName}_${enzymeName}_e-0_g-0_a_s-${lowerSize}_S-${upperSize}.txt)
		    echo -e "more details in file stdout_extract_fragments_${genomeName}_${enzymeName}_e-0_g-0_a_s-${lowerSize}_S-${upperSize}.txt"
	    fi
	  fi
	  if $cleanTmp; then
	    rm -f ${tmpPrefix}.bed.gz
	  fi
	  
	  # step 4 ------------------------------------------------------------------
	  if hash R 2>/dev/null; then
	    if [ $verbose -gt "0" ]; then
		    echo -e "plot histogram of kept fragment sizes (with R)..."
	    fi
	    cmd="x <- read.table(\"${tmpPrefix}_frags.bed.gz\", sep=\"\t\"); tmp <- x[,3] - x[,2]; pdf(\"hist_frags_${genomeName}_${enzymeName}_e-0_g-0_a_s-${lowerSize}_S-${upperSize}.pdf\"); hist(tmp[tmp <= quantile(tmp, 0.75)], main=\"${enzymeName} fragments on ${genomeName}\", xlab=\"fragment size (bp)\", breaks=\"FD\"); abline(v=${lowerSize}, lwd=2, col=\"red\"); abline(v=${upperSize}, lwd=2, col=\"red\"); dev.off()"
	    echo $cmd | R --vanilla --quiet >/dev/null 2>&1
	  fi
	  
  fi # nbCutSites > 0
}

verbose=1
genomeFile=""
genomeName=""
enzymeFile=""
enzymeName=""
lowerSize="100"
upperSize="300"
useFullHeader="no"
cleanTmp=false
pathToInsilicut=""
parseCmdLine "$@"

if [ $verbose -gt "0" ]; then
  startTime=$(timer)
  msg="START ${0##*/} ${progVersion} $(date +"%Y-%m-%d") $(date +"%H:%M:%S")"
  msg+="\ncmd-line: $0 "$@ # comment if an option takes a glob as argument
  msg+="\ncwd: $(pwd)"
  echo -e $msg
fi

run genomeFile genomeName enzymeFile enzymeName lowerSize upperSize useFullHeader cleanTmp pathToInsilicut verbose

if [ $verbose -gt "0" ]; then
  msg="END ${0##*/} ${progVersion} $(date +"%Y-%m-%d") $(date +"%H:%M:%S")"
  msg+=" ($(timer startTime))"
  echo $msg
fi
